test/sharness/t0082-repo-gc-auto.sh

```diff
@@ -0,0 +1,56 @@
+#!/bin/sh
+#
+# MIT Licensed; see the LICENSE file in this repository.
+#
+
+test_description="Test ipfs repo auto gc"
+
+. lib/test-lib.sh
+
+test_init_ipfs
+
+test_expect_success "generate 2 600 kB files and 2 MB file using go-random" '
+    random 600k 41 >600k1 &&
+    random 600k 42 >600k2 &&
+    random 2M 43 >2M
+'
+
+test_expect_success "set ipfs gc watermark, storage max, and gc timeout" '
+    test_config_set Datastore.StorageMax "2MB" &&
+    test_config_set --json Datastore.StorageGCWatermark 60 &&
+    test_config_set Datastore.GCPeriod "20ms"
+'
+
+test_launch_ipfs_daemon
+
+test_expect_success "adding data below watermark doesn't trigger auto gc" '
+    ipfs add 600k1 >/dev/null &&
+    disk_usage "$IPFS_PATH/blocks" >expected &&
+    go-sleep 40ms &&
+    disk_usage "$IPFS_PATH/blocks" >actual &&
+    test_cmp expected actual
+'
+
+#todo: `ipfs pin rm -r` is unclean so cant use du -sb "$IPFS_PATH/blocks" | awk "{print \$1}" >expected &&
+test_expect_success "adding data beyond watermark triggers auto gc" '
+    if test $(uname -s) = "Darwin"; then
+        expected=1000
+    else
+        expected=1000000
+    fi &&
+    HASH=`ipfs add -q 600k2` &&
+    ipfs pin rm -r $HASH &&
+    go-sleep 40ms &&
+    actual=$(disk_usage "$IPFS_PATH/blocks") &&
+    test "$actual" -lt "$expected"
+'
+
+test_expect_success "adding data beyond storageMax should not work" '
+    echo "Error: file size exceeds slack space allowed by storageMax. Maybe unpin some files?" >expected &&
+    test_must_fail ipfs add 2M 2>actual &&
+    test_cmp expected actual
+'
+
```

Comment:
we may also want some stress tests. like a loop version of:

```
for ...
  random= ...  a random size < watermark, so it takes a few iterations to cross threshold
  hash=$(random <seed> $random | ipfs add -q | tail -n1)
  ipfs pin rm -r $hash
  test_storage_size_below_storage_max
```

and loop enough times adding enough content that gc must have been triggered many times (say ~50). this is just good to stress the repeated operation.

(we probably need a `ipfs add --pinned=false` :) )